arch/xtensa/include/xtensa_asm2_context.h

```diff
@@ -106,6 +106,18 @@ struct xtensa_irq_base_save_area {
 	uintptr_t fpu15;
 #endif
 
+#if (XCHAL_CP_MASK & (1 << XCHAL_CP_ID_AUDIOENGINELX)) && \
+	defined(CONFIG_CPU_HAS_DSP) && defined(CONFIG_DSP_SHARING)
+
+
+	/*
+	 * Carve space for the registers used by the audio engine coprocessor
+	 * (which is always CP1). Carve additional space to manage alignment
+	 * at run-time.
+	 */
+	uint8_t  ae_regs[XCHAL_CP1_SA_SIZE + XCHAL_CP1_SA_ALIGN];
+#endif
```

Comment:
Any particular reason why this is located here ? Can it end up leaving a hole in the struct depending on XCHAL_CP1_SA_SIZE + XCHAL_CP1_SA_ALIGN ?

Reply:
Assuming the first question refers to why it was placed in the `struct xtensa_irq_base_save_area` -- it seemed like a reasonable place to store the AudioEngine HiFi registers as the floating point registers were already here.

As to the potential hole ... there is definitely going to be some unused space. When the HiFi3 or HiFi4 is enabled, XCHAL_CP1_SA_ALIGN evaluates to 8. However, from what I can tell, `struct xtensa_irq_base_save_area` is not guaranteed to be aligned to an 8-byte boundary (see its use in `struct xtensa_irq_stack_frame_aX` where X = {3, 7, 11, 15}). To work around this, the alignment has been added to the size to provide extra space so that we can force the necessary alignment at runtime.

Reply:
Sorry, repeating from elsewhere:  Indeed, the BSA isn't aligned in any special way on entry, but there's no reason that couldn't be the case.  Just put the ae_regs[] array at an aligned spot within this struct, and align the stack before pushing the registers.

Reply:
For what it is worth, initial simple attempts to just align the structure via the` __align()` macro are not sufficient (resulting in fatal exceptions). This suggests that there is assembly code that carves out space for the BSA that results in an unaligned BSA. Subsequent searches for a generated offset that would provide further evidence has turned up `___xtensa_irq_bsa_t_SIZEOF`, which is indeed used in the assembly files to carve space (or at least adjust the pointers). I am currently digging into what exactly would need to force the correct alignment when those pointers are updated.